### PR TITLE
Compartmental models in NESTML

### PR DESCRIPTION
**Pull request by LeanderEwert**
Generalized ion-channels. Adding concentration mechanisms (work in progress).

To do
------
- [x] Finish processing review comments on this PR
- [x] Render and possibly fix documentation
- [x] Fix ODE-toolbox and do a new release (see https://github.com/nest/ode-toolbox/pull/73)
- [x] Fix CI

### CODE REVIEW

**Comment by pnbabu** [comment]
@LeanderEwert along with your changed files, there seem to be other files (unrelated to your PR) modified in this PR. Could you please check and update it?

**Comment by clinssen** [comment]
Replaces https://github.com/nest/nestml/pull/772.

**Review by clinssen — CHANGES_REQUESTED** [review]
Could you please fix the static code checks? See https://github.com/nest/nestml/actions/runs/5069718027/jobs/9376337549?pr=872

File: .gitignore [hunk 1]
```diff
@@ -26,3 +27,14 @@ venv
 *.gdf
 *~
 *.iml
+/generated/
```
Comment (clinssen) [inline .gitignore:30]: Changes not relevant to this PR, please revert to original .gitignore.

File: doc/fig/nestml_clip_art.png
Comment (clinssen) [inline doc/fig/nestml_clip_art.png:1]: This file was removed upstream, please merge master and make sure the figure is gone from the PR.

File: pynestml/cocos/co_co_convolve_cond_correctly_built.py [hunk 2]
```diff
@@ -38,13 +38,13 @@ class CoCoConvolveCondCorrectlyBuilt(CoCo):
     """
 
     @classmethod
-    def check_co_co(cls, node):
+    def check_co_co(cls, neuron):
```
Comment (clinssen) [inline pynestml/cocos/co_co_convolve_cond_correctly_built.py:41]: Unnecessary change, please remove from the PR. Same for the files below.

**Review by clinssen — CHANGES_REQUESTED** [review]

File: models/cm_default.nestml [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    tau_r_AN_AMPA real = 0.2 ms      # Synaptic Time Constant Excitatory Synapse
+    tau_d_AN_AMPA real = 3.0 ms      # Synaptic Time Constant Excitatory Synapse
+    e_AN_NMDA real = 0 mV            # NMDA reversal Potential
+    tau_r_AN_NMDA real = 0.2 ms      # Synaptic Time Constant NMDA Synapse
+    tau_d_AN_NMDA real = 43.0 ms     # Synaptic Time Constant NMDA Synapse
+    NMDA_ratio real = 2.0      # NMDA_ratio
+  end
+
+  equations:
+    """
```
Comment (clinssen) [inline models/cm_default.nestml:72]: Please change this into a multiline comment prepended by ``#`` on each line, rather than ``"""``.

**Review by clinssen — CHANGES_REQUESTED** [review]

**Comment by clinssen** [comment]
N.B.: make sure to credit Krzysztof Blaszczyk as an author when this is merged.

File: pynestml/utils/ast_utils.py [hunk 4]
```diff
@@ -1769,20 +1776,26 @@ def replace_inline_expressions_through_defining_expressions(cls, definitions: Se
...
-                target.rhs = ModelParser.parse_expression(target_definition)
-                target.update_scope(m.get_scope())
-                target.accept(ASTSymbolTableVisitor())
-
-                def log_set_source_position(node):
-                    if node.get_source_position().is_added_source_position():
-                        node.set_source_position(source_position)
-
-                target.accept(ASTHigherOrderVisitor(visit_funcs=log_set_source_position))
+            if "mechanism" not in [e.namespace for e in m.get_decorators()]:
```
Comment (clinssen) [inline pynestml/utils/ast_utils.py:1779]: Could you move this filtering step to the place where the function is called? Then this function does not have to change.

File: pynestml/utils/chan_info_enricher.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+from pynestml.utils.model_parser import ModelParser
+from pynestml.visitors.ast_symbol_table_visitor import ASTSymbolTableVisitor
+import sympy
+
+from pynestml.utils.mechs_info_enricher import MechsInfoEnricher
+
+
+class ChanInfoEnricher(MechsInfoEnricher):
+    """
+    Class extends MechanismInfoEnricher by the computation of the inline derivative. This hasn't been done in the
```
Comment (clinssen) [inline pynestml/utils/chan_info_enricher.py:31]:
```suggestion
    Class extends MechsInfoEnricher by the computation of the inline derivative. This hasn't been done in the
```

File: pynestml/utils/channel_processing.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                    cls.check_if_key_zero_var_for_expression(str(sympy_expression.args[0]), var_str) and \
+                    cls.check_if_key_zero_var_for_expression(str(sympy_expression.args[1]), var_str):
+                return True
+            elif isinstance(sympy_expression, sympy.core.mul.Mul) and \
+                    (cls.check_if_key_zero_var_for_expression(str(sympy_expression.args[0]), var_str) or \
+                    cls.check_if_key_zero_var_for_expression(str(sympy_expression.args[1]), var_str)):
+                return True
+            elif rhs_expression_str == var_str:
+                return True
+            else:
```
Comment (clinssen) [inline pynestml/utils/channel_processing.py:61]: you can use the "return early" pattern here and eliminate most "else" statements.

File: pynestml/utils/channel_processing.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        """
+        Searching for parameters in the root-expression that if zero lead to the expression always being zero so that
+        the computation may be skipped.
+        """
+        mechs_info = cls.write_key_zero_parameters_for_root_inlines(mechs_info)
+
+        return mechs_info
+
+    @classmethod
+    def check_if_key_zero_var_for_expression(cls, rhs_expression_str, var_str):
```
Comment (clinssen) [inline pynestml/utils/channel_processing.py:48]: A brief docstring here and there about the purpose of a (helper) function would be great for code documentation.

File: pynestml/utils/logger.py [hunk 8]
```diff
@@ -143,8 +145,18 @@ def log_message(cls, node: ASTNode = None, code: MessageCode = None, message: st
         if cls.no_print:
             return
         if cls.logging_level.value <= log_level.value:
+            node_name = ''
+            if isinstance(node, ASTInlineExpression):
+                node_name = node.variable_name
+            # elif isinstance (node, ASTInputPort):
```
Comment (clinssen) [inline pynestml/utils/logger.py:151]: remove

File: pynestml/utils/mechanism_processing.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+        return mechs_info
+
+    @classmethod
+    def ode_toolbox_processing(cls, neuron, mechs_info):
+        mechs_info = cls.prepare_equations_for_ode_toolbox(neuron, mechs_info)
+        mechs_info = cls.collect_raw_odetoolbox_output(mechs_info)
+        return mechs_info
+
+    def collect_information_for_specific_mech_types(cls, neuron, mechs_info):
```
Comment (clinssen) [inline pynestml/utils/mechanism_processing.py:124]: Should be marked @classmethod?

File: pynestml/utils/mechs_info_enricher.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+
+                        solution_transformed["propagators"][variable_name] = {
+                            "ASTVariable": prop_variable, "init_expression": expression, }
+                        expression_variable_collector = ASTEnricherInfoCollectorVisitor()
+                        expression.accept(expression_variable_collector)
+
+                        neuron_internal_declaration_collector = ASTEnricherInfoCollectorVisitor()
+                        neuron.accept(neuron_internal_declaration_collector)
+
+                        # print("TRV: " + PredefinedFunctions.TIME_RESOLUTION)
```
Comment (clinssen) [inline pynestml/utils/mechs_info_enricher.py:112]: remove

File: pynestml/utils/mechs_info_enricher.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+                        for variable in expression_variable_collector.all_variables:
+                            for internal_declaration in neuron_internal_declaration_collector.internal_declarations:
+                                # print(internal_declaration.get_variables()[0].get_name())
```
Comment (clinssen) [inline pynestml/utils/mechs_info_enricher.py:115]: remove; you can also use logging debug print for text that could be meaningful to the user

File: pynestml/utils/messages.py [hunk 12]
```diff
@@ -162,10 +176,10 @@ def get_lexer_error(cls):
         message = 'Error occurred during lexing: abort'
         return MessageCode.LEXER_ERROR, message
 
-    @classmethod
-    def get_could_not_determine_cond_based(cls, type_str, name):
-        message = "Unable to determine based on type '" + type_str + "' of variable '" + name + "' whether conductance-based or current-based"
-        return MessageCode.LEXER_ERROR, message
+    # @classmethod
```
Comment (clinssen) [inline pynestml/utils/messages.py:179]: why comment out?

File: tests/docstring_comment_test.py [hunk 13]
```diff
@@ -57,15 +66,21 @@ class DocstringCommentTest(unittest.TestCase):
     def test_docstring_success(self):
         self.run_docstring_test('valid')
 
-    @pytest.mark.xfail(strict=True, raises=ParseCancellationException)
+    # for some reason xfail is completely ignored when executing on my machine (python 3.8.5)
```
Comment (clinssen) [inline tests/docstring_comment_test.py:69]: comment still relevant?

File: tests/invalid/CoCoCmFunctionExists.nestml [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+neuron cm_model_one_invalid:
+
+  state:
+    # indicate that we have a compartmental model
+    # by declaring an unused variable with the name "v_comp"
+    v_comp real = 0.0
+    
+    m_Na real = 0.0
+    
+  end
```
Comment (clinssen) [inline tests/invalid/CoCoCmFunctionExists.nestml:45]: end keyword

**Review by clinssen — COMMENTED** [review]
Thanks a lot for the great work! I mostly just have some small nitpicks.

One thing I am missing from the PR is documentation. It would be great to have a new .rst file under "running NESTML", and a new link next to NEST and the Python targets. This could say a few words about the new features and how to use them (perhaps it could contain some sample code borrowed from the compartmental_models test).

File: tests/invalid/CoCoCmFunctionOneArg.nestml [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+  state:
+    # indicate that we have a compartmental model
+    # by declaring an unused variable with the name "v_comp"
+    v_comp real = 0.0
+    
+    m_Na real = 0.0
+    h_Na real = 0.0
+    
+  end
```
Comment (pnbabu) [inline tests/invalid/CoCoCmFunctionOneArg.nestml:46]: Please remove `end` keyword from all the .nestml files.

**Review by pnbabu — CHANGES_REQUESTED** [review]

File: extras/convert_cm_default_to_template.py
[the remark this answers is not in the document]
Comment (LeanderEwert) [reply extras/convert_cm_default_to_template.py:1]: I actually don't know. I think this is Willem's creation. It might help convert a prototype/manually written model in c++ into a jinja template.

File: models/cm_default.nestml [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+
+Author
+++++++
+Willem Wybo
+"""
+neuron cm_default:
+
+    state:
+
+        # the presence of the state variable [v_comp]
```
[the remark this answers is not in the document]
Comment (LeanderEwert) [reply models/cm_default.nestml:24]: Hey the description was indeed outdated but as far as I see the state must still be defined so that the parser does not complain that the variable in the equations can't be resolved.

File: pynestml/cocos/co_co_compartmental_model.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+
+from pynestml.cocos.co_co import CoCo
+from pynestml.meta_model.ast_neuron import ASTNeuron
+from pynestml.utils.channel_processing import ChannelProcessing
+
+
+class CoCoCompartmentalModel(CoCo):
+    @classmethod
+    def check_co_co(cls, neuron: ASTNeuron):
+        """
```
[the remark this answers is not in the document]
Comment (LeanderEwert) [reply pynestml/cocos/co_co_compartmental_model.py:30]: Yes, thank you. I just fixed that. Also the naming scheme of those cm related coco functions and classes didn't make much sense. Also changed that.

File: pynestml/codegeneration/printers/nest_variable_printer.py [hunk 18]
```diff
@@ -102,7 +102,8 @@ def print_variable(self, variable: ASTVariable) -> str:
 
         if symbol.is_inline_expression:
             # there might not be a corresponding defined state variable; insist on calling the getter function
-            return "get_" + self._print(variable, symbol, with_origin=False) + vector_param + "()"
+            # return "get_" + self._print(variable, symbol, with_origin=False) + vector_param + "()"
+            return self._print(variable, symbol, with_origin=False) #temporary modification to not enforce getter function
```
[the remark this answers is not in the document]
Comment (LeanderEwert) [reply pynestml/codegeneration/printers/nest_variable_printer.py]: I have added a member variable that can be set via the constructor which controls this behavior now.

File: pynestml/codegeneration/resources_nest_compartmental/cm_neuron/@NEURON_NAME@.cpp.jinja2 [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::{{neuronSpecificFileNamesCmSyns["main"]}}::{{neuronSpecificFileNamesCmSyns["main"]}}()
+  : ArchivingNode()
+  , c_tree_()
+  , syn_buffers_( 0 )
+  , logger_( *this )
+  , V_th_( -55.0 )
```
[the remark this answers is not in the document]
Comment (LeanderEwert) [reply pynestml/codegeneration/resources_nest_compartmental/cm_neuron/@NEURON_NAME@.cpp.jinja2:49]: it can be changed in the simulation setup. But I agree that it would be nice.

File: pynestml/codegeneration/resources_nest_compartmental/cm_neuron/@NEURON_NAME@.h.jinja2 [hunk 20]
```diff
@@ -0,0 +1,343 @@
...
+namespace nest
+{
+
+/* BeginUserDocs: neuron, compartmental model
+
+Short description
++++++++++++++++++
+
+A neuron model with user-defined dendrite structure.
+Currently, AMPA, GABA or AMPA+NMDA receptors.
```
[the remark this answers is not in the document]
Comment (LeanderEwert) [reply pynestml/codegeneration/resources_nest_compartmental/cm_neuron/@NEURON_NAME@.h.jinja2:44]: Haven't looked to close at this yet but looks to me like a leftover from prototype to template conversion.

File: pynestml/codegeneration/resources_nest_compartmental/cm_neuron/directives
[the remark this answers is not in the document]
Comment (LeanderEwert) [reply pynestml/codegeneration/resources_nest_compartmental/cm_neuron/directives:1]: it's a link to point_neuron/directives which is needed at least with the way that the compartmental templates are accessing these files located there right now.

File: pynestml/utils/channel_processing.py (hunk 6 above)
Comment (LeanderEwert) [reply pynestml/utils/channel_processing.py:61]: Don't you think that this would make it very hard to read?

File: pynestml/utils/synapse_processing.py [hunk 21]
```diff
@@ -0,0 +1,232 @@
...
+                          kernel_buffer):
+        kernel_buffers = {tuple(kernel_buffer)}
+        odetoolbox_indict = cls.transform_ode_and_kernels_to_json(
+            neuron, parameters_block, kernel_buffers)
+        odetoolbox_indict["options"] = {}
+        odetoolbox_indict["options"]["output_timestep_symbol"] = "__h"
+        return odetoolbox_indict
+
+    @classmethod
+    def transform_ode_and_kernels_to_json(
```
[the remark this answers is not in the document]
Comment (LeanderEwert) [reply pynestml/utils/synapse_processing.py:163]: I didn't write it but it seems like a modification of that function that is very specific to the processing context of this class. Are you sure you would like it to be moved?

File: extras/codeanalysis/check_copyright_headers.py
Comment (pnbabu) [inline extras/codeanalysis/check_copyright_headers.py:1]: Why is this file empty?

**Review by pnbabu — CHANGES_REQUESTED** [review]

File: extras/convert_cm_default_to_template.py
[the remark this answers is not in the document]
Comment (clinssen) [reply extras/convert_cm_default_to_template.py:1]: cc @WillemWybo

[the remark this answers is not in the document]
Comment (WillemWybo) [reply extras/convert_cm_default_to_template.py:1]: This script converts the generic parts (`cm_default.*` and `cm_tree.*`) of the default compartmental model in NEST to a .jinja template. It is therefor a helper tool for developers working concurrently on the compartmental models in NEST and NESTML. It should however be used with extreme caution, as it doesn't automatically update the compartmentcurrents. 

Is it ok to add the header and a comment  along the aforementioned lines, and leave the script in its current location?

**Review by clinssen — APPROVED** [review]

**Review by pnbabu — APPROVED** [review]

**Review by WillemWybo — APPROVED** [review]

**Comment by clinssen** [comment]
Thanks to everyone who contributed to this!

NESTML 7.0.0 will be released soon, containing this feature.
